base/inference.jl

```diff
@@ -542,7 +542,38 @@ add_tfunc(===, 2, 2,
         end
         return Bool
     end)
-add_tfunc(isdefined, 1, IInf, (args...)->Bool)
+function isdefined_tfunc(args...)
+    arg1 = args[1]
+    if isa(arg1, Const)
+        a1 = typeof(arg1.val)
+    else
+        a1 = widenconst(arg1)
+    end
+    if isType(a1)
+        return Bool
+    end
+    if isleaftype(a1)
+        if a1 <: Array # TODO
+        elseif a1 === Module # TODO
+        elseif length(args) == 2 && isa(args[2], Const)
+            n = nfields(a1)
+            val = args[2].val
+            idx::Int = 0
+            if isa(val, Symbol)
+                idx = fieldindex(a1, val, false)
+            elseif isa(val, Int)
+                idx = val::Int
```

Comment:
type assert is redundant with branch on isa Int and declaration of `idx` as being of type Int